src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -1007,7 +1017,20 @@ static inline bool notification_match(reservation_t& res,
         return ret;
       }
     }
-    res.topics.emplace_back(topic_filter.s3_id, topic_cfg, res_id);
+    // load the topic,if there is change in topic config while it's stored in
+    // notification.
+    rgw_pubsub_topic result;
+    const RGWPubSub ps(res.store, res.user_tenant);
+    auto ret = ps.get_topic(res.dpp, topic_cfg.name, result, res.yield);
+    if (ret < 0) {
```

Comment:
not sure about the list, i would prefer to keep that only for telemetry/debug.
would prefer to go with lazy cleanup - if the notification is not used after topic deletion, there is not harm in the fact that the attribute is still o nthe bucket.
regarding v1, there is no cleanup process, and, IMO, we don't need to fix that

Reply:
> not sure about the list, i would prefer to keep that only for telemetry/debug.
> would prefer to go with lazy cleanup - if the notification is not used after topic deletion, there is not harm in the fact that the attribute is still o nthe bucket.
> regarding v1, there is no cleanup process, and, IMO, we don't need to fix that

I feel since the list is anyways going to be there, so while deleting the topic it's just about reading the list and deleting the notifications? What could be wrong there ?

Reply:
the problem is regarding the mechanism that maintain the list.
if this is only for telemetry, it has different requirements that if it is for the actual functionality

Reply:
> the problem is regarding the mechanism that maintain the list.
> if this is only for telemetry, it has different requirements that if it is for the actual functionality

Mechanism ?? Originally didn't we mention 2 major reasons for holding bucket mapping. 1>. Delete/update bucket notification when topic is updated/deleted (update is not problem anymore, as we are reading queue from disc,  but deletion is still pending so we can use the list. And 2> telemetry

Reply:
careful about deleting bucket notifications if the topic isn't available. metadata sync doesn't make any guarantees about ordering, so it may replicate changes to bucket metadata before its associated topic metadata, and you could see ENOENT here in the meantime

Reply:
> careful about deleting bucket notifications if the topic isn't available. metadata sync doesn't make any guarantees about ordering, so it may replicate changes to bucket metadata before its associated topic metadata, and you could see ENOENT here in the meantime

This is extreme rare scenario, topic and bucket notifications are created on primary, and only notification has synced on secondary n also on primary there is a object out which has synced to secondary  before topic has synced and that sync object creates a notification event and this code path does not find the topic and ends up deleting the notification entry...
Doesn't look realistic ?

Reply:
consider a client workload continuously writing objects to a bucket on a secondary zone

on primary, create a topic and add a notification policy to that bucket

if the bucket metadata replicates to the secondary first, some of those object writes may see the notification xattr but fail to find the associated topic. trying to delete the notification xattr in that case would be wrong

but writing bucket metadata should be conditional on `is_meta_master()` anyway, and that seems to avoid any issues here

Reply:
> consider a client workload continuously writing objects to a bucket on a secondary zone
> 
> on primary, create a topic and add a notification policy to that bucket
> 
> if the bucket metadata replicates to the secondary first, some of those object writes may see the notification xattr but fail to find the associated topic. trying to delete the notification xattr in that case would be wrong
> 
> but writing bucket metadata should be conditional on `is_meta_master()` anyway, and that seems to avoid any issues here

@yuvalif with possibility of race condition from above scenario, thinking of not adding the code to delete the topic in this code path. anycase deleting the topic here was a stop gap if topic was deleted but notification did not. ? thoughts

Reply:
agree. this will be too risky, and we won't be able to recover from it